### ISSUE TITLE
Question about the usage of VHAP in GaussianAvatars

### ISSUE THREAD

**Issue by justin4ai** [ShenhanQian/VHAP#19]
Hello, first of all I really appreciate you provide such a wonderful project to the world! I'm learning a lot from your works and they are very helpful since the paper and corresponding code are well structured :)

However, it is difficult for me to understand what VHAP does, especially in GaussianAvatars. As far as I understand, GaussianAvatars train Gaussians built on the coordinates of the tracked FLAME mesh, which means the mesh is ready-made (or complete in advance before the beginning of the training) and not learned during the training.

Since VHAP seemingly has no technical report, I'd like to ask you about the mechanism of VHAP - I am not able to watch through a viewer and even run your code for some reasons: 

1. Is VHAP needed to be fitted or trained to a specific subject(e.g. NeRSemble dataset with a certain identity) then usable? Or is it a ready-made generalized mesh tracker that I can infer a mesh which has exactly the same form to one in GaussianAvatars demo .gif?

2. Is my understand that the mesh is frozen during the training of GaussianAvatars once it is tracked via VHAP in the beginning correct?

I'm still learning from yours so my question even might be lack of background of graphics but I would greatly appreciate your help if you give me a response with this question!

Best,
Junyeong Ahn

**Comment by ShenhanQian** [comment]
Hi Junyeong, thanks for your questions.

1. VHAP is based on the FLAME model, which is generic in identities and not limited to a specific dataset.
2. GaussianAvatars relies on the tracked FLAME parameters for the best quality. However, finetuning the FLAME parameters with GaussianAvatars further improves visual details, as shown in our paper.

**Comment by justin4ai** [comment]
@ShenhanQian Appreciate your kind explanation!

1. Honestly it is hard to imagine how it works and I cannot find the part VHAP is used in GaussianAvatars ```train.py```. I'd like to learn by myself as mush as possible rather than pestering you without any effort, so would you kindly point the line of code or the file in GaussianAvatars where VHAP is used to track the mesh?

2. Yes I found FLAME parameters become learnable after [initializing](https://github.com/ShenhanQian/GaussianAvatars/blob/main/scene/flame_gaussian_model.py#L174), but I'm still confused with that FlameHead class in FlameGaussianModel class loads the canonical ,,template'' mesh(FLAME_MESH_PATH = "flame_model/assets/flame/head_template_mesh.obj"), not a canonical personalized mesh (e.g. [personalized mesh for 306](https://github.com/ShenhanQian/GaussianAvatars/blob/main/media/demo.gif)). So I wonder if such offsets between the template and the personalized mesh are learned while that _fine-tuning_ or I'm missing something?

Best,
Junyeong Ahn

**Comment by Jp-17** [comment]
Here's my understanding, hope it helps.

1. VHAP is not directly used within GaussianAvatars, but rather it's used as a preprocessing step to obtain FLAME parameters which are then packaged and used in GaussianAvatars (via the -s parameter in python train.py). You can find the code for packaging the FLAME parameters after VHAP processing here: https://github.com/ShenhanQian/VHAP/blob/main/jobs/combine_nersemble.sh

2. Regarding the template mesh - since VHAP provides FLAME head parameters (what you mean "offsets between the template and the personalized mesh"), these are loaded and used in GaussianAvatars (see https://github.com/ShenhanQian/GaussianAvatars/blob/ce6474b75ca866651cb74fb8129e278538d084bc/scene/flame_gaussian_model.py#L43) to compute the personalized mesh from the canonical template mesh. After this, the 3D gaussian points are bound to this mesh and optimized.

**Comment by justin4ai** [comment]
@Jp-17 Thanks for your help!!! I looked through and would like to ask you as follows:

1. Oh so you mean ```data/UNION10_${SUBJECT}_EMO1234EXP234589_v16_DS2-0.5x_lmkSTAR_teethV3_SMOOTH_offsetS_whiteBg_maskBelowLine``` is already processed by VHAP and originally not provided by NeRSemble? I'm not given the official dataset (still waiting after signing) so I didn't know. 

2. If yes for my understand for 1.,  are FLAMEHead parameters  (= offsets from the template to the personalized mesh which are contained in ```data/UNION10_${SUBJECT}_EMO1234EXP234589_v16_DS2-0.5x_lmkSTAR_teethV3_SMOOTH_offsetS_whiteBg_maskBelowLine``` processed by VHAP) used as ```tgt_train_meshes``` in ```load_meshes``` function?

If my questions are unclear, please let me know. Sincerely appreciate your help!!

**Comment by Jp-17** [comment]
1. Yes

2. Yes, but it's not just for tgt_train_meshes. The VHAP-processed data in data/UNION10_*** contains FLAME parameters under different conditions. These parameters serve multiple purposes:
- Some are used for training (train_meshes)
- Some for testing (test_meshes)
- And some for driving the trained 3DGS model (tgt_train_meshes)
You can check this link (https://github.com/ShenhanQian/GaussianAvatars/blob/ce6474b75ca866651cb74fb8129e278538d084bc/scene/dataset_readers.py#L297) to understand how the dataset is split

**Comment by justin4ai** [comment]
@Jp-17 Learned a lot from your help thanks a lot! Now I somewhat understand. Actually I still don't understand the differences among train_meshes/test_meshes/tgt_train_meshes/tgt_test_meshes exactly despite your kind explanation :( At least the question about VHAP is enough answered by you two guys thanks! I realize most of information related to flame including transformations must be ones tracked by VHAP - ```UNION10_${SUBJECT}_EMO1234EXP234589_v16_DS2-0.5x_lmkSTAR_teethV3_SMOOTH_offsetS_whiteBg_maskBelowLine``` is just an aggregated version and other folders are also already processed with VHAP I think?

I would ask more questions after reviewing more and organizing my knowledge. Many thanks!!

**Comment by justin4ai** [comment]
Hi, I'd like to ask a question about the usage of UV-mapping in GaussianAvatars although this is not related to VHAP for your convenience.

In [GaussianAvatars - FlameHead class](https://github.com/ShenhanQian/GaussianAvatars/blob/main/flame_model/flame.py#L154) , there are some code lines for 
 uv-mapping:

```python
        verts, faces, aux = load_obj(flame_template_mesh_path, load_textures=False)

        vertex_uvs = aux.verts_uvs
        face_uvs_idx = faces.textures_idx  # index into verts_uvs

        # create uvcoords per face --> this is what you can use for uv map rendering
        # range from -1 to 1 (-1, -1) = left top; (+1, +1) = right bottom
        # pad 1 to the end
        pad = torch.ones(vertex_uvs.shape[0], 1)
        vertex_uvs = torch.cat([vertex_uvs, pad], dim=-1)
        vertex_uvs = vertex_uvs * 2 - 1
        vertex_uvs[..., 1] = -vertex_uvs[..., 1]

        face_uv_coords = face_vertices(vertex_uvs[None], face_uvs_idx[None])[0]
        self.register_buffer("face_uvcoords", face_uv_coords, persistent=False)
        self.register_buffer("faces", faces.verts_idx, persistent=False)

        self.register_buffer("verts_uvs", aux.verts_uvs, persistent=False)
        self.register_buffer("textures_idx", faces.textures_idx, persistent=False)
```

, which I don't understand their explicit usage of them in further steps. For which part do you specifically need uv-mapping? Do GaussianAvatars then have any part of reverse step (= from uv to 3D)?

Also, can I ask about the role of ```FlameMask class```? I see it is used in [some case](https://github.com/ShenhanQian/GaussianAvatars/blob/main/scene/flame_gaussian_model.py#L264) to filter from/under ```self.binding```, but I find it hard to understand the exact role even through struggling in code level.


Sorry for asking many questions about details because of my lack of knowledge, but I'll be so glad if you help me out with my questions. I did a lot of investigation and of course will keep doing before you answer! Thanks for your inspiring project.

Best,
Junyeong Ahn

**Comment by ShenhanQian** [comment]
Hi Junyeong,

UV-mapping and region-masking are heavily relied on by VHAP but not actually used by GaussianAvatars. I didn't remove them thinking that they may be useful for others for future work.

**Comment by justin4ai** [comment]
Oh I see I found FlameUvMask class over [here](https://github.com/ShenhanQian/VHAP/blob/main/vhap/model/flame.py#L1057) and it literally contains masks for each region in ```FLAME_UVMASK_PATH = "asset/flame/uv_masks.npz"``` which is supposed to be defined in the same UV space you construct for template mesh in ```FlameHead class```, right?

Then would you let me know how uv-mapping in ```FlameHead class``` is used in VHAP? As you said, your implementation of uv-mapping would be very useful for future work so I'd like to learn from and refer to it a lot. Probably the most important part is how it is used (e.g. mapping back from uv to 3D, manipulation in uv space and so on) but I have difficulty finding them. I'd be so glad if you let me know the role of it in VHAP case!

Your work is awesome, thanks!
